ports/esp32/mpthreadport.c

```diff
@@ -59,14 +59,22 @@ STATIC thread_t *thread = NULL; // root pointer, handled by mp_thread_gc_others
 void mp_thread_init(void *stack, uint32_t stack_len) {
     mp_thread_set_state(&mp_state_ctx.thread);
     // create the first entry in the linked list of all threads
-    thread = &thread_entry0;
-    thread->id = xTaskGetCurrentTaskHandle();
-    thread->ready = 1;
-    thread->arg = NULL;
-    thread->stack = stack;
-    thread->stack_len = stack_len;
-    thread->next = NULL;
+    thread_entry0.id = xTaskGetCurrentTaskHandle();
+    thread_entry0.ready = 1;
+    thread_entry0.arg = NULL;
+    thread_entry0.stack = stack;
+    thread_entry0.stack_len = stack_len;
+    thread_entry0.next = NULL;
     mp_thread_mutex_init(&thread_mutex);
+
+    // memory barrier
+    #if CONFIG_IDF_TARGET_ESP32
+    asm volatile ("memw" ::: "memory");
+    #else
+    asm volatile ("" ::: "memory");
```

Comment:
I think this should be `__sync_synchronize();`, which would work for xtensa as well, ie all variants.

Otherwise for RISC (esp32-c3) it's "fence".